README.md
# blox

A Clojure library designed to create parametric 3d models. 

It's inspired by https://github.com/farrellm/scad-clj but the emphasis is on composable functions rather than macros. 

## Usage

The simplest way to try this out is to execute the example functions tx0 .. tx4 towards the end of the scad.clj file. These generate .scad files in the out directory which can be viewed with the Openscad software available from [OpenSCAD.org](http://www.openscad.org/index.html). Or you can just take a look at these pictures:

t0 - an odd shape with some translucent parts:

![t0.clj - an odd shape with some translucent parts](out/t0.png)

t1 - a sphere with a cylinder subtracted:

![t1 - a sphere with a cylinder subtracted](out/t1.png)

t2 - Jackstone?

![t2 - - Jackstone?](out/t2.png)

t3 - the interior of a Jackstone:

![t3](out/t3.png)

t4 - I have no idea what to call this, it's a hull around a scaled jackstone:

![t4](out/t4.png)

The other file, brick.clj, is an example of using the convenience functions (currently limited to cube, cylinder, difference, translate and union). The brick and plate functions generate something that looks like a lego component. Please note that a 3d printed version of this is almost certainly not going to actually be lego compatible. The interior of the brick is not quite right and the measurements are off. 

(brick 3 4) - an impossible brick:

![3x4 exterior](out/3x4 exterior.png)

(brick 3 4) - a view of the interior:

![3x4 interior](out/3x4 interior.png)



## License

Copyright A. Dwelly © 2017

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-01-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-01-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/blox/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/blox/compare/0.1.0...0.1.1


test/blox/core_test.clj
(ns blox.core-test
  (:require [clojure.test :refer :all]
            [blox.scad :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/blox/core.clj
(ns blox.core
  (:require [clojure.string :as s]))

(defn name-of [key]
  (if (or (= key :fn) (= key :fa) (= key :fs))
    (str "$" (name key))
    (name key)))

(defn val-of [val]
  (if (or (list? val) (vector? val))
    (str "[" (s/join "," val) "]")
    (str val)))

(defn func-to-str [{:keys [vec] :as p}]
  (let [params (for [[k v] (dissoc (:params p) :trans)] (str (name-of k) "=" (val-of v)))
        param-str (s/join "," params)
        vc (when vec (val-of vec))]
    (str (if (get-in p [:params :trans]) (str "#" (:name p)) (:name p)) "(" (if vc vc "") param-str ")")))

(defn add-param [{:keys [params] :as shape} k v]
  (assoc shape :params (assoc params k v)))

(defn func
  ([name, params]
   {:name name :params params})
  ([name, params, defaults]
   {:name name :params (merge defaults params)})
  ([name vec params defaults]
   {:name name :vec vec :params params :defaults defaults}))

(def cylinder-defaults {:h 20 :center true :fa 0.5 :fs 0.5})
(def cube-defaults {:size [20 30 40] :center true})
(def sphere-defaults {:r 20 :center true})

(defmulti primitive-gen :primitive)

(defmethod primitive-gen :sphere [{:keys [r center trans]}]
  (let [mandatory (func "sphere" {} sphere-defaults)]
    (cond-> mandatory
            trans (add-param :trans trans)
            r (add-param :r r)
            (some? center) (add-param :center center))))

(defmethod primitive-gen :cube [{:keys [size center trans] :as shape}]
  (let [mandatory
        (if size
          (func "cube" {:size size} cube-defaults)
          (func "cube" {} cube-defaults))]
    (cond-> mandatory
            trans (add-param :trans trans)
            (some? center) (add-param :center center))))

(defmethod primitive-gen :cylinder [{:keys [r r1 r2 h center fa fs fn trans] :as shape}]
  (let [mandatory
        (cond (and r r1 r2) (println "ERROR: cylinder contains r, r1, and r2 " shape)
              (and r1 r2) (func "cylinder" {:r1 r1 :r2 r2} cylinder-defaults)
              r (func "cylinder" {:r r} cylinder-defaults)
              :else (func "cylinder" {:r 20} cylinder-defaults))]
    (cond-> mandatory
            h (add-param :h h)
            fa (add-param :fa fa)
            fs (add-param :fs fs)
            fn (add-param :fn fn)
            trans (add-param :trans trans)
            (some? center) (add-param :center center))))    ;center is boolean

(defmulti operator-gen :operator)

(defmethod operator-gen :color [{:keys [vec]}] (func "color" vec {} {}))

(defmethod operator-gen :difference [_] (func "difference" {} {}))

(defmethod operator-gen :hull [_] (func "hull" {} {}))

(defmethod operator-gen :intersection [_] (func "intersection" {} {}))

(defmethod operator-gen :rotate [{:keys [vec]}] (func "rotate" vec {} {}))

(defmethod operator-gen :scale [{:keys [vec]}] (func "scale" vec {} {}))

(defmethod operator-gen :translate [{:keys [vec]}] (func "translate" vec {} {}))

(defmethod operator-gen :union [_] (func "union" {} {}))

(defn code-gen [shape]
  (cond (contains? shape :primitive) (-> shape primitive-gen func-to-str (str ";\n"))
        (contains? shape :operator) (let [sub-shapes (map code-gen (:content shape))
                                          sub-str (s/join sub-shapes)]
                                      (str (-> shape operator-gen func-to-str) "{\n" sub-str "}\n"))
        :else (println "Unknown shape " shape)))

(defn write [file shape]
  (spit file (code-gen shape)))

;; Examples

;; A coloured arbitrary object
(defn ex0 [] (write "out/t0.scad" {:operator :color
                                   :vec      [0, 0, 1]
                                   :content  [{:operator :union
                                               :content  [{:primitive :cylinder :r 10 :h 55}
                                                          {:primitive :sphere :trans true}
                                                          {:operator :translate
                                                           :vec      [5, 5, 5]
                                                           :content  [{:primitive :cube :size [20 30 40] :trans true}]}]}]}))

;; A sphere with a hole (punched out with a translucent cylinder
(defn ex1 [] (write "out/t1.scad"
                    {:operator :difference :content [{:primitive :sphere}
                                                     {:primitive :cylinder :r 10 :h 50 :trans true}]}))

;; 3d cross
(defn ex2 [] (write "out/t2.scad"
                    {:operator :union :content [{:operator :rotate :vec [90, 0, 0] :content [{:primitive :cylinder :r 10 :h 50}]}
                                                {:operator :rotate :vec [0, 90, 0] :content [{:primitive :cylinder :r 10 :h 50}]}
                                                {:operator :rotate :vec [0, 0, 90] :content [{:primitive :cylinder :r 10 :h 50}]}]}))

  ; Smooth polyhedron with a circular profile in x, y, z
  (defn ex3 [] (write "out/t3.scad"
                      {:operator :intersection
                       :content  [{:operator :rotate
                                   :vec      [90, 0, 0] :content [{:primitive :cylinder
                                                                   :r         10
                                                                   :h         50}]}
                                  {:operator :rotate
                                   :vec      [0, 90, 0]
                                   :content  [{:primitive :cylinder :r 10 :h 50}]}
                                  {:operator :rotate
                                   :vec      [0, 0, 90]
                                   :content  [{:primitive :cylinder :r 10 :h 50}]}]}))

  ;; Weird spaceship/submarine shape
  (defn ex4 [] (write "out/t4.scad"
                      {:operator :hull
                       :content  [{:operator :scale
                                   :vec      [3, 1, 1]
                                   :content  [{:operator :rotate
                                               :vec      [90, 0, 0]
                                               :content  [{:primitive :cylinder
                                                           :r         10
                                                           :h         50}]}
                                              {:operator :rotate
                                               :vec      [0, 90, 0]
                                               :content  [{:primitive :cylinder :r 10 :h 50}]}
                                              {:operator :rotate
                                               :vec      [0, 0, 90]
                                               :content  [{:primitive :cylinder :r 10 :h 50}]}]}]}))



src/blox/brick.clj
(ns blox.brick
  (:require [blox.scad :refer [write cube cylinder difference translate union]]))

(def iota 0.2)
(def brick-dim 7.8)
(def half-brick-dim (* brick-dim 0.5))
(def tile-height 3.21)
(def brick-height (* tile-height 3))
(def half-brick-height (* brick-height 0.5))
(def stud-height 1.77)
(def half-stud-height (* stud-height 0.5))
(def stud-radius 2.45)
(def wall-depth 1)
(def buttress-height 5)
(def half-buttress-height (* buttress-height 0.5))
(def buttress-width 0.84)
(def tube-height (- brick-height iota))
(def tube-radius 3.25)

(defn studded-cube [l w h stud-h stud-r]
  {:operator :color
   :vec      [1 0 0]
   :content  [(union (cube l w h)
                     (translate [0 0 half-stud-height] (cylinder (+ h stud-h) stud-r)))]})

(defn stud1 [h]
  (studded-cube brick-dim brick-dim h stud-height stud-radius))

(defn solid [h m n]
  (let [row (for [i (range m)] (translate [(* i brick-dim) 0 0] (stud1 h)))
        cols (for [j (range n)] (translate [0 (* j brick-dim) 0] (union row)))]
    (translate [(+ half-brick-dim (* -1 half-brick-dim m)) (+ half-brick-dim (* -1 half-brick-dim n)) 0] (union cols))))


(defn buttress [studs]
  (cube buttress-width (- (* studs brick-dim) iota) buttress-height))

(defn tube []
  (difference (cylinder tube-height tube-radius)
              (cylinder (+ tube-height iota) (- tube-radius buttress-width))))

(defn interior [m n]
  (let [half-total-width (* m half-brick-dim)
        half-total-length (* n half-brick-dim)
        buttresses (for [i (range (dec m))] (translate [(- (* (inc i) brick-dim) half-total-width) 0 (- half-brick-height half-buttress-height iota)] (buttress n)))
        tubes (for [i (range (dec m)) j (range (dec n))] (translate[(- (* (inc i) brick-dim) half-total-width) (- (* (inc j) brick-dim) half-total-length) 0] (tube)))]
    (union (concat buttresses tubes))))

(defn plate [m n]
  (solid tile-height m n))

(defn brick [m n]
  (union (interior m n)
         (difference (solid brick-height m n)
                     (translate [0 0 (* -1 wall-depth)] (cube (- (* m brick-dim) wall-depth) (- (* n brick-dim) wall-depth) brick-height)))))

(defn make []
  (write "out/stud1" (brick 3 4)))

src/blox/scad.clj
(ns blox.scad
  (:require [clojure.string :as s]))

(defn name-of [key]
  (if (or (= key :fn) (= key :fa) (= key :fs))
    (str "$" (name key))
    (name key)))

(defn val-of [val]
  (cond (or (list? val) (vector? val)) (str "[" (s/join "," (map val-of val)) "]")
        (float? val) (format "%.2f" val)
        :else (str val)))

(defn func-to-str [{:keys [name params vec] :as p}]
  (let [pms (for [[k v] (dissoc params :trans)] (str (name-of k) "=" (val-of v)))
        param-str (s/join "," pms)
        vc (when vec (val-of vec))]
    (str (if (:trans params) (str "#" name) name) "(" (if vc vc "") param-str ")")))

(defn add-param [{:keys [params] :as shape} k v]
  (assoc shape :params (assoc params k v)))

(defn func
  ([name, params]
   {:name name :params params})
  ([name, params, defaults]
   {:name name :params (merge defaults params)})
  ([name vec params defaults]
   {:name name :vec vec :params params :defaults defaults}))

(def cylinder-defaults {:h 20 :center true :fa 0.5 :fs 0.5})
(def cube-defaults {:size [20 30 40] :center true})
(def sphere-defaults {:r 20 :center true})

(defmulti primitive-gen :primitive)

(defmethod primitive-gen :sphere [{:keys [r center trans]}]
  (let [mandatory (func "sphere" {} sphere-defaults)]
    (cond-> mandatory
            trans (add-param :trans trans)
            r (add-param :r r)
            (some? center) (add-param :center center))))

(defmethod primitive-gen :cube [{:keys [size center trans] :as shape}]
  (let [mandatory
        (if size
          (func "cube" {:size size} cube-defaults)
          (func "cube" {} cube-defaults))]
    (cond-> mandatory
            trans (add-param :trans trans)
            (some? center) (add-param :center center))))

(defmethod primitive-gen :cylinder [{:keys [r r1 r2 h center fa fs fn trans] :as shape}]
  (let [mandatory
        (cond (and r r1 r2) (println "ERROR: cylinder contains r, r1, and r2 " shape)
              (and r1 r2) (func "cylinder" {:r1 r1 :r2 r2} cylinder-defaults)
              r (func "cylinder" {:r r} cylinder-defaults)
              :else (func "cylinder" {:r 20} cylinder-defaults))]
    (cond-> mandatory
            h (add-param :h h)
            fa (add-param :fa fa)
            fs (add-param :fs fs)
            fn (add-param :fn fn)
            trans (add-param :trans trans)
            (some? center) (add-param :center center))))    ;center is boolean

(defmulti operator-gen :operator)

(defmethod operator-gen :color [{:keys [vec]}] (func "color" vec {} {}))

(defmethod operator-gen :difference [_] (func "difference" {} {}))

(defmethod operator-gen :hull [_] (func "hull" {} {}))

(defmethod operator-gen :intersection [_] (func "intersection" {} {}))

(defmethod operator-gen :rotate [{:keys [vec]}] (func "rotate" vec {} {}))

(defmethod operator-gen :scale [{:keys [vec]}] (func "scale" vec {} {}))

(defmethod operator-gen :translate [{:keys [vec]}] (func "translate" vec {} {}))

(defmethod operator-gen :union [_] (func "union" {} {}))

;;(defmethod operator-gen :module [_] {:keys [params]} (func "module" {:params params})

(defn code-gen [shape]
  (cond (contains? shape :primitive) (-> shape primitive-gen func-to-str (str ";\n"))
        (contains? shape :operator) (let [sub-shapes (map code-gen (:content shape))
                                          sub-str (s/join sub-shapes)]
                                      (str (-> shape operator-gen func-to-str) "{\n" sub-str "}\n"))
        :else (println "Unknown shape " shape)))

(defn write [file shape]
  (let [file-name (if (s/ends-with? file ".scad") file (str file ".scad"))]
    (spit file-name (code-gen shape))))

;; Primitive Examples

;; A coloured arbitrary object
(defn ex0 [] (write "out/t0.scad" {:operator :color
                                   :vec      [0, 0, 1]
                                   :content  [{:operator :union
                                               :content  [{:primitive :cylinder :r 10 :h 55}
                                                          {:primitive :sphere :trans true}
                                                          {:operator :translate
                                                           :vec      [5, 5, 5]
                                                           :content  [{:primitive :cube :size [20 30 40] :trans true}]}]}]}))

;; A sphere with a hole (punched out with a translucent cylinder
(defn ex1 [] (write "out/t1.scad"
                    {:operator :difference :content [{:primitive :sphere}
                                                     {:primitive :cylinder :r 10 :h 50 :trans true}]}))

;; 3d cross
(defn ex2 [] (write "out/t2.scad"
                    {:operator :union :content [{:operator :rotate :vec [90, 0, 0] :content [{:primitive :cylinder :r 10 :h 50}]}
                                                {:operator :rotate :vec [0, 90, 0] :content [{:primitive :cylinder :r 10 :h 50}]}
                                                {:operator :rotate :vec [0, 0, 90] :content [{:primitive :cylinder :r 10 :h 50}]}]}))

  ; Smooth polyhedron with a circular profile in x, y, z
  (defn ex3 [] (write "out/t3.scad"
                      {:operator :intersection
                       :content  [{:operator :rotate
                                   :vec      [90, 0, 0] :content [{:primitive :cylinder
                                                                   :r         10
                                                                   :h         50}]}
                                  {:operator :rotate
                                   :vec      [0, 90, 0]
                                   :content  [{:primitive :cylinder :r 10 :h 50}]}
                                  {:operator :rotate
                                   :vec      [0, 0, 90]
                                   :content  [{:primitive :cylinder :r 10 :h 50}]}]}))

  ;; Weird spaceship/submarine shape
  (defn ex4 [] (write "out/t4.scad"
                      {:operator :hull
                       :content  [{:operator :scale
                                   :vec      [3, 1, 1]
                                   :content  [{:operator :rotate
                                               :vec      [90, 0, 0]
                                               :content  [{:primitive :cylinder
                                                           :r         10
                                                           :h         50}]}
                                              {:operator :rotate
                                               :vec      [0, 90, 0]
                                               :content  [{:primitive :cylinder :r 10 :h 50}]}
                                              {:operator :rotate
                                               :vec      [0, 0, 90]
                                               :content  [{:primitive :cylinder :r 10 :h 50}]}]}]}))

;; Convenience functions

(defn cube [x y z]
  {:primitive :cube
   :size [x y z]})

(defn cylinder [h r]
  {:primitive :cylinder
   :h h
   :r r})

(defn difference [a b & shapes]
  {:operator :difference
   :content (concat [a b] shapes)})

(defn translate [v & shapes]
  {:operator :translate
   :vec v
   :content (vec shapes)})

(defn union
  ([a] {:operator :union :content (if (seq? a) a [a])})
  ([a b & shapes]
    {:operator :union
     :content (concat [a b] shapes)}))

